Add ConfigLoader and TemplatedConfigLoader deprecation notices on documentation. (#2991)
kedro-org/kedro: docs/source/configuration/advanced_configuration.md

```

`OmegaConfigLoader` can load `YAML` and `JSON` files. Acceptable file extensions are `.yml`, `.yaml`, and `.json`. By default, any configuration files used by the config loaders in Kedro are `.yml` files.

To use `OmegaConfigLoader` in your project, set the `CONFIG_LOADER_CLASS` constant in your [`src/<package_name>/settings.py`](../kedro_project_setup/settings.md):

```python
from kedro.config import OmegaConfigLoader  # new import

CONFIG_LOADER_CLASS = OmegaConfigLoader
```
### Advanced `OmegaConfigLoader` features
Some advanced use cases of `OmegaConfigLoader` are listed below:
- [How to do templating with the `OmegaConfigLoader`](#how-to-do-templating-with-the-omegaconfigloader)
- [How to use global variables with the `OmegaConfigLoader`](#how-to-use-global-variables-with-the-omegaconfigloader)
- [How to use resolvers in the `OmegaConfigLoader`](#how-to-use-resolvers-in-the-omegaconfigloader)
- [How to load credentials through environment variables](#how-to-load-credentials-through-environment-variables)

## TemplatedConfigLoader

> ```{note}

Reviewer: Can we make this a warning instead?

```suggestion
```{warning}
```

By default, Kedro is set up to use the [ConfigLoader](/kedro.config.ConfigLoader) class. Kedro also provides two additional configuration loaders with more advanced functionality: the [TemplatedConfigLoader](/kedro.config.TemplatedConfigLoader) and the [OmegaConfigLoader](/kedro.config.OmegaConfigLoader).
Each of these classes are alternatives for the default `ConfigLoader` and have different features. The following sections describe each of these classes and their specific functionality in more detail.

## OmegaConfigLoader

> [OmegaConf](https://omegaconf.readthedocs.io/) is a Python library tailored for managing settings. It operates as a YAML-based hierarchical system for organizing configurations and supports merging configurations from various sources.

Reviewer 2: ```suggestion
[OmegaConf](https://omegaconf.readthedocs.io/) is a Python library designed to handle and manage settings. It serves as a YAML-based hierarchical system to organise configurations, which can be structured to accommodate various sources, allowing you to merge settings from multiple locations.
```

From Kedro 0.18.5 you can use the [`OmegaConfigLoader`](/kedro.config.OmegaConfigLoader) which uses `OmegaConf` to load data.

```{note}

> `OmegaConfigLoader` is under active development. It was first available from Kedro 0.18.5 with additional features due in later releases. Let us know if you have any feedback about the `OmegaConfigLoader`.

Reviewer 2: ```suggestion
`OmegaConfigLoader` is under active development. It is available from Kedro version 0.18.5 with additional features due in later releases. Let us know if you have any feedback about the `OmegaConfigLoader`.
```

Reviewer 2: You need to say how to let us know (either add a link to Slack or a mailto link)

```

`OmegaConfigLoader` can load `YAML` and `JSON` files. Acceptable file extensions are `.yml`, `.yaml`, and `.json`. By default, any configuration files used by the config loaders in Kedro are `.yml` files.

To use `OmegaConfigLoader` in your project, set the `CONFIG_LOADER_CLASS` constant in your [`src/<package_name>/settings.py`](../kedro_project_setup/settings.md):

```python
from kedro.config import OmegaConfigLoader  # new import

CONFIG_LOADER_CLASS = OmegaConfigLoader
```
### Advanced `OmegaConfigLoader` features
Some advanced use cases of `OmegaConfigLoader` are listed below:

> - [Advanced configuration](#advanced-configuration)
>   - [OmegaConfigLoader](#omegaconfigloader)
>     - [Advanced `OmegaConfigLoader` features](#advanced-omegaconfigloader-features)
>   - [TemplatedConfigLoader](#templatedconfigloader)
>     - [Provide template values through globals](#provide-template-values-through-globals)
>   - [Advanced Kedro configuration](#advanced-kedro-configuration)
>     - [How to change which configuration files are loaded](#how-to-change-which-configuration-files-are-loaded)
>     - [How to ensure non default configuration files get loaded](#how-to-ensure-non-default-configuration-files-get-loaded)
>     - [How to bypass the configuration loading rules](#how-to-bypass-the-configuration-loading-rules)
>     - [How to use Jinja2 syntax in configuration](#how-to-use-jinja2-syntax-in-configuration)
>     - [How to do templating with the `OmegaConfigLoader`](#how-to-do-templating-with-the-omegaconfigloader)
>       - [Parameters](#parameters)
>       - [Catalog](#catalog)
>       - [Other configuration files](#other-configuration-files)
>     - [How to use global variables with the `OmegaConfigLoader`](#how-to-use-global-variables-with-the-omegaconfigloader)
>     - [How to use resolvers in the `OmegaConfigLoader`](#how-to-use-resolvers-in-the-omegaconfigloader)
>     - [How to load credentials through environment variables](#how-to-load-credentials-through-environment-variables)

Reviewer 3: ```suggestion
- [How to do templating with the `OmegaConfigLoader`](#how-to-do-templating-with-the-omegaconfigloader)
- [How to use global variables with the `OmegaConfigLoader`](#how-to-use-global-variables-with-the-omegaconfigloader)
- [How to use resolvers in the `OmegaConfigLoader`](#how-to-use-resolvers-in-the-omegaconfigloader)
- [How to load credentials through environment variables](#how-to-load-credentials-through-environment-variables)
```

Reviewer 3: Only link sections relevant to advanced `OmegaConfigLoader` usage, not all subheadings

## TemplatedConfigLoader

```{note}

> `ConfigLoader` and `TemplatedConfigLoader` have been deprecated since Kedro `0.18.12` and will be removed in Kedro `0.19.0`. Refer to the [Migration guide for config loaders](./config_loader_migration.md) for instructions on how to update your code base to use `OmegaConfigLoader`.

Reviewer 2: ```suggestion
`ConfigLoader` and `TemplatedConfigLoader` have been deprecated since Kedro `0.18.12` and will be removed in Kedro `0.19.0`. Refer to the [migration guide for config loaders](./config_loader_migration.md) for instructions on how to update your code to use `OmegaConfigLoader`.
```

If you specify both `globals_pattern` and `globals_dict` in `CONFIG_LOADER_ARGS`, the contents of the dictionary resulting from `globals_pattern` are merged with the `globals_dict` dictionary. In case of conflicts, the keys from the `globals_dict` dictionary take precedence.

## Advanced Kedro configuration

This section contains a set of guidance for advanced configuration requirements of standard Kedro projects:

> - [How to change which configuration files are loaded](#how-to-change-which-configuration-files-are-loaded)

Reviewer: You might want to undo these changes as well, @stichbury do we have a style guide on Markdown? I think we're always using `*` for bulleted lists but correct me if I'm wrong.

Reviewer 2: I've seen both and don't mind either. I use `*` myself.

If you specify both `globals_pattern` and `globals_dict` in `CONFIG_LOADER_ARGS`, the contents of the dictionary resulting from `globals_pattern` are merged with the `globals_dict` dictionary. In case of conflicts, the keys from the `globals_dict` dictionary take precedence.

## Advanced Kedro configuration

This section contains a set of guidance for advanced configuration requirements of standard Kedro projects:

> - [Advanced configuration](#advanced-configuration)
>   - [OmegaConfigLoader](#omegaconfigloader)
>     - [Advanced `OmegaConfigLoader` features](#advanced-omegaconfigloader-features)
>   - [TemplatedConfigLoader](#templatedconfigloader)
>     - [Provide template values through globals](#provide-template-values-through-globals)
>   - [Advanced Kedro configuration](#advanced-kedro-configuration)
>     - [How to change which configuration files are loaded](#how-to-change-which-configuration-files-are-loaded)
>     - [How to ensure non default configuration files get loaded](#how-to-ensure-non-default-configuration-files-get-loaded)
>     - [How to bypass the configuration loading rules](#how-to-bypass-the-configuration-loading-rules)
>     - [How to use Jinja2 syntax in configuration](#how-to-use-jinja2-syntax-in-configuration)
>     - [How to do templating with the `OmegaConfigLoader`](#how-to-do-templating-with-the-omegaconfigloader)
>       - [Parameters](#parameters)
>       - [Catalog](#catalog)
>       - [Other configuration files](#other-configuration-files)
>     - [How to use global variables with the `OmegaConfigLoader`](#how-to-use-global-variables-with-the-omegaconfigloader)
>     - [How to use resolvers in the `OmegaConfigLoader`](#how-to-use-resolvers-in-the-omegaconfigloader)
>     - [How to load credentials through environment variables](#how-to-load-credentials-through-environment-variables)

Reviewer 3: ```suggestion
- [How to change which configuration files are loaded](#how-to-change-which-configuration-files-are-loaded)
- [How to ensure non default configuration files get loaded](#how-to-ensure-non-default-configuration-files-get-loaded)
- [How to bypass the configuration loading rules](#how-to-bypass-the-configuration-loading-rules)
- [How to use Jinja2 syntax in configuration](#how-to-use-jinja2-syntax-in-configuration)
- [How to do templating with the `OmegaConfigLoader`](#how-to-do-templating-with-the-omegaconfigloader)
 - [How to use global variables with the `OmegaConfigLoader`](#how-to-use-global-variables-with-the-omegaconfigloader)
- [How to use resolvers in the `OmegaConfigLoader`](#how-to-use-resolvers-in-the-omegaconfigloader)
- [How to load credentials through environment variables](#how-to-load-credentials-through-environment-variables)
```

Reviewer 3: This also is linking to previous sections that we don't need to mention here.

Reviewer: I suspect this is the Markdown VSCode extension? I've observed similar things locally. Maybe the `toctree` directive from Sphinx would be useful here too.

Author: It was indeed the extension. Should be fixed now.
